src/blockchain/ethereum.rs

```diff
@@ -41,119 +46,234 @@ impl Ethereum {
+    /// future into a reactor.
+    pub fn stream_blocks(&self) -> impl Stream<Item = Block, Error = Error> {
+        // Blocks filter is a future that returns a filter.
+        let blocks_filter = self.web3.eth_filter().create_blocks_filter();
+
+        // Block hashes is a stream of block hashes.
+        let block_hashes = blocks_filter
+            .map(|filter| filter.stream(Duration::from_secs(1)))
+            .flatten_stream();
+
+        // Web3 blocks is a stream of block futures, mapped from a stream of block hashes.
+        let web3_clone = self.web3.clone();
+        let web3_blocks = block_hashes
+            .map_err(|error| {
+                Error::new(
+                    ErrorKind::NodeError,
+                    format!("Error while streaming blocks from node: {}", error),
+                )
+            }).and_then(move |block_hash| {
+                web3_clone
+                    .eth()
+                    .block(BlockId::from(block_hash))
+                    .map_err(|error| {
+                        Error::new(
+                            ErrorKind::NodeError,
+                            format!("Was not able to retrieve block: {}", error),
+                        )
+                    })
+            });
 
-        Ok(ethereum)
+        // Returns a stream of blocks, mapped from a stream of web3 block futures.
+        web3_blocks.and_then(|web3_block| match web3_block {
+            // Mapping web3 block Option to a Block.
+            // Wrapping in Ok() as it has to return an IntoFuture.
+            Some(web3_block) => match web3_block.into_block() {
+                Ok(block) => Ok(block),
+                Err(error) => Err(Error::new(
+                    ErrorKind::NodeError,
+                    format!("Could not convert block from web3: {}", error),
+                )),
+            },
+            None => Err(Error::new(
+                ErrorKind::NodeError,
+                "No block found".to_string(),
+            )),
+        })
     }
 
     /// Uses web3 to retrieve the accounts.
     /// Converts them to blockchain addresses and returns all addresses in a
     /// vector.
-    pub fn get_accounts(&self) -> Vec<Address> {
-        let addresses = self.web3.eth().accounts().wait().unwrap();
-        let mut v = Vec::new();
-
-        for h160 in addresses {
-            match h160.as_address() {
-                Ok(address) => v.push(address),
-                Err(error) => warn!("Unable to convert h160 to address: {}", error),
-            }
-        }
+    pub fn get_accounts(&self) -> impl Future<Item = Vec<Address>, Error = Error> {
+        self.web3
+            .eth()
+            .accounts()
+            .map_err(|error| {
+                Error::new(
+                    ErrorKind::NodeError,
+                    format!("Was not able to retrieve accounts: {}", error),
+                )
+            }).and_then(|addresses| {
```

Comment:
shouldn't this be `map` instead of `and_then`?
similar to `sign` method.

Reply:
True, I *think.* I will update it.